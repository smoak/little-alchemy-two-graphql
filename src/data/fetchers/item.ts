import low from 'lowdb';
import FileSync from 'lowdb/adapters/FileSync';

const adapter = new FileSync<DatabaseSchema>('db.json');
const db = low(adapter);

interface DatabaseSchema {
  readonly items: DatabaseItems;
}

export interface DatabaseItem {
  readonly myths: boolean;
  readonly makes: Record<string, string>;
  readonly combinations: string[][];
}

type DatabaseItems = Record<string, DatabaseItem>;

type GetItems = () => DatabaseItems;
export const getItems: GetItems = () => db.get('items').value();

type GetItemByName = (name: string) => DatabaseItem | undefined;
export const getItemByName: GetItemByName = name => db.get('items').value()[name];

type GetItemNames = () => string[];
export const getItemNames: GetItemNames = () => db.get('items').keys().value();
