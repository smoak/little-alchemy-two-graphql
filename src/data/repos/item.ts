import Fuse from 'fuse.js';

import { DatabaseItem, getItemByName, getItemNames, getItems } from '../fetchers/item';

export interface Item {
  readonly name: string;
  readonly myths: boolean;
  readonly creates: ItemCombination[];
  readonly combinations: ItemCombination[];
  readonly imageUrl: string;
}

export interface ItemCombination {
  readonly source: string;
  readonly target: string;
}

const fromCombinations = (combinations: DatabaseItem['combinations']): ItemCombination[] =>
  combinations.map(combo => ({ source: combo[0], target: combo[1] }));

const fromItemWithName = (name: string, item: DatabaseItem): Item => ({
  name,
  imageUrl: item.image_url,
  myths: item.myths,
  creates: item.makes,
  combinations: fromCombinations(item.combinations),
});

type FindById = (id: string) => Item;
export const findById: FindById = id => {
  const item = getItemByName(id);

  if (!item) {
    throw new Error(`Could not find item with id: ${id}`);
  }

  return fromItemWithName(id, item);
};

type FindAll = () => Item[];
export const findAll: FindAll = () => {
  const items = getItems();

  return Object.keys(items).map<Item>(itemName => {
    const item = items[itemName];
    return fromItemWithName(itemName, item);
  });
};

type Search = (query: string) => Item[];
export const search: Search = query => {
  const items = getItemNames();
  const options: Fuse.IFuseOptions<string> = {
    minMatchCharLength: 2,
    threshold: 1,
  };
  const fuse = new Fuse(items, options);
  const results = fuse.search(query);
  return results.map(result => findById(result.item));
};
