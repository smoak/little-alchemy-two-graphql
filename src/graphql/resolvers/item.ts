import { GraphQLFieldResolver } from 'graphql';

import { ItemCombination, findAll, findById } from '../../data/repos/item';
import { Query } from '../types';

export const ItemsResolver: GraphQLFieldResolver<Query, unknown, unknown> = (_source, _args, _context, _info) =>
  findAll();

interface ItemWithNameResolverArgs {
  readonly name: string;
}
export const ItemWithNameResolver: GraphQLFieldResolver<Query, unknown, ItemWithNameResolverArgs> = (_source, args) =>
  findById(args.name);

export const SourceItemCombinationResolver: GraphQLFieldResolver<ItemCombination, unknown, unknown> = (
  parent,
  _args,
  _context,
  _info
) => findById(parent.source);

export const TargetItemCombinationResolver: GraphQLFieldResolver<ItemCombination, unknown, unknown> = (
  parent,
  _args,
  _context,
  _info
) => findById(parent.target);
