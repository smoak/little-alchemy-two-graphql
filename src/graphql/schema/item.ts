import {
  GraphQLBoolean,
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
  Thunk,
} from 'graphql';
import { globalIdField } from 'graphql-relay';

import { Item, ItemCombination } from '../../data/repos/item';
import { SourceItemCombinationResolver, TargetItemCombinationResolver } from '../resolvers';

import { nodeInterface } from './node';

interface ItemFields extends GraphQLFieldConfigMap<Item, unknown, unknown> {
  readonly name: GraphQLFieldConfig<Item, unknown, unknown>;
  readonly myths: GraphQLFieldConfig<Item, unknown, unknown>;
  readonly id: GraphQLFieldConfig<unknown, unknown, unknown>;
  readonly creates: GraphQLFieldConfig<Item, unknown, unknown>;
  readonly combinations: GraphQLFieldConfig<Item, unknown, unknown>;
}

interface ItemCombinationFields extends GraphQLFieldConfigMap<ItemCombination, unknown, unknown> {
  readonly source: GraphQLFieldConfig<ItemCombination, unknown, unknown>;
  readonly target: GraphQLFieldConfig<ItemCombination, unknown, unknown>;
}

type Fields = () => ItemFields;
const fields: Fields = () => ({
  name: { type: new GraphQLNonNull(GraphQLString) },
  myths: { type: new GraphQLNonNull(GraphQLBoolean) },
  id: globalIdField('Item', obj => obj.name),
  creates: {
    description: 'The items that this item creates',
    type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(ItemCombinationType))),
  },
  combinations: {
    description: 'The items that create this item',
    type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(ItemCombinationType))),
  },
});

const itemCombinationFields: Thunk<ItemCombinationFields> = () => ({
  source: { type: new GraphQLNonNull(ItemType), resolve: SourceItemCombinationResolver },
  target: { type: new GraphQLNonNull(ItemType), resolve: TargetItemCombinationResolver },
});

export const ItemType = new GraphQLObjectType<Item, unknown, unknown>({
  name: 'Item',
  description: 'An item',
  interfaces: [nodeInterface],
  fields,
  isTypeOf: source =>
    Object.keys(source).includes('name') &&
    Object.keys(source).includes('myths') &&
    Object.keys(source).includes('creates'),
});

export const ItemCombinationType = new GraphQLObjectType<ItemCombination, unknown, unknown>({
  name: 'ItemCombination',
  description: 'A combination of two items',
  fields: itemCombinationFields,
});
