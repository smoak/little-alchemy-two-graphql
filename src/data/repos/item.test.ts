jest.mock('../fetchers/item');

import { DatabaseItem, getItemByName, getItems } from '../fetchers/item';

import { findAll, findById } from './item';

describe('item', () => {
  const item: DatabaseItem = {
    combinations: [['item1', 'item2']],
    makes: {
      item1: 'item2',
    },
    myths: false,
  };

  describe('.findById', () => {
    let result: ReturnType<typeof findById>;

    describe('when the item cannot be found', () => {
      beforeEach(() => {
        (getItemByName as jest.Mock).mockReturnValue(undefined);

        result = findById('foo');
      });

      it('is `undefined`', () => {
        expect(result).toBeUndefined();
      });
    });

    describe('when the item can be found', () => {
      beforeEach(() => {
        (getItemByName as jest.Mock).mockReturnValue(item);

        result = findById('item');
      });

      it('returns the expected result', () => {
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('.findAll', () => {
    let result: ReturnType<typeof findAll>;

    describe('when there are no items', () => {
      beforeEach(() => {
        (getItems as jest.Mock).mockReturnValue([]);

        result = findAll();
      });

      it('retuns the expected result', () => {
        expect(result).toEqual([]);
      });
    });

    describe('when there are items', () => {
      beforeEach(() => {
        const items = {
          ['item']: item,
        };
        (getItems as jest.Mock).mockReturnValue(items);

        result = findAll();
      });

      it('returns the expected result', () => {
        expect(result).toMatchSnapshot();
      });
    });
  });
});
