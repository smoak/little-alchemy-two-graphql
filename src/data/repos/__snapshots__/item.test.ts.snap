// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`item .findAll when there are items returns the expected result 1`] = `
Array [
  Object {
    "combinations": Array [
      Object {
        "source": "item1",
        "target": "item2",
      },
    ],
    "creates": Array [
      Object {
        "source": "item1",
        "target": "item2",
      },
    ],
    "myths": false,
    "name": "item",
  },
]
`;

exports[`item .findById when the item can be found returns the expected result 1`] = `
Object {
  "combinations": Array [
    Object {
      "source": "item1",
      "target": "item2",
    },
  ],
  "creates": Array [
    Object {
      "source": "item1",
      "target": "item2",
    },
  ],
  "myths": false,
  "name": "item",
}
`;
